sklearn/manifold/t_sne.py

```diff
@@ -706,6 +706,8 @@ def _fit(self, X, skip_num_points=0):
                             'the array is small enough for it to fit in '
                             'memory. Otherwise consider dimensionality '
                             'reduction techniques (e.g. TruncatedSVD)')
+        if self.method == 'barnes_hut':
+            X = check_array(X, ensure_min_samples=2, dtype=np.float64)
```

Comment:
why shouldn't barnes_hut support sparse matrices?

Reply:
because of the previous if statement?

```python
if self.method == 'barnes_hut' and sp.issparse(X):
             raise TypeError('A sparse matrix was passed, but dense '
                             'data is required for method="barnes_hut"
```